src/server/bitops_family.cc

```diff
@@ -57,14 +64,93 @@ constexpr int32_t GetByteIndex(uint32_t offset) noexcept {
 }
 
+constexpr std::uint8_t CountBitsRange(std::uint8_t byte, std::uint8_t from, uint8_t to) {
+  int count = 0;
+  for (int i = from; i < to; i++) {
+    count += CheckBitStatus(byte, GetNormalizedBitIndex(i));
+  }
+  return count;
+}
+
+// Count the number of bits that are on, on bytes boundaries: i.e. Start and end are the indices for
+// bytes locations inside str CountBitSetByByteIndices
+std::size_t CountBitSetByByteIndices(std::string_view at, std::size_t start, std::size_t end) {
+  if (start >= end) {
+    return 0;
+  }
+  end = std::min(end, at.size());  // don't overflow
+  std::uint32_t count =
+      std::accumulate(std::next(at.begin(), start), std::next(at.begin(), end), 0,
+                      [](auto counter, uint8_t ch) { return counter + absl::popcount(ch); });
+  return count;
+}
+
+// Count the number of bits that are on, on bits boundaries: i.e. Start and end are the indices for
+// bits locations inside str
+std::size_t CountBitSetByBitIndices(std::string_view at, std::size_t start, std::size_t end) {
+  auto first_byte_index = GetByteIndex(start);
+  auto last_byte_index = GetByteIndex(end);
+  if (start % OFFSET_FACTOR == 0 && end % OFFSET_FACTOR == 0) {
+    return CountBitSetByByteIndices(at, first_byte_index, last_byte_index);
+  }
+  const auto last_bit_first_byte =
+      first_byte_index != last_byte_index ? OFFSET_FACTOR : GetBitIndex(end);
+  const auto first_byte = GetByteValue(at, start);
+  std::uint32_t count = CountBitsRange(first_byte, GetBitIndex(start), last_bit_first_byte);
+  if (first_byte_index < last_byte_index) {
+    first_byte_index++;
+    const auto last_byte = GetByteValue(at, end);
+    count += CountBitsRange(last_byte, 0, GetBitIndex(end));
+    count += CountBitSetByByteIndices(at, first_byte_index, last_byte_index);
+  }
+  return count;
+}
+
+// General purpose function to count the number of bits that are on.
+// The parameters for start, end and bits are defaulted to the start of the string,
+// end of the string and bits are false.
+// Note that when bits is false, it means that we are looking on byte boundaries.
+std::size_t CountBitSet(std::string_view str, int64_t start, int64_t end, bool bits) {
+  const int32_t size = bits ? str.size() * OFFSET_FACTOR : str.size();
+
+  auto NormalizedOffset = [size](int32_t orig) {
+    if (orig < 0) {
+      orig = size + orig;
+    }
+    return orig;
+  };
+
+  if (start > 0 && end > 0 && end < start) {
+    return 0;  // for illegal range with positive we just return 0
+  }
+
+  if (start < 0 && end < 0 && start > end) {
+    return 0;  // for illegal range with negative we just return 0
+  }
+
+  start = NormalizedOffset(start);
+  if (end > 0 && end < start) {
+    return 0;
+  }
+  end = NormalizedOffset(end);
+  if (start > end) {
+    std::swap(start, end);  // we're going backward
+  }
+  if (end > size) {
+    end = size;  // don't overflow
+  }
+  return bits ? CountBitSetByBitIndices(str, start, ++end)
```

Comment:
i am sorry to be a nudnik, by doing increment inside ternary operator when passing to an argument makes me nervous.
can you please pull out `++end` to the line before?

Reply:
sure